### PR TITLE
feat(api): add API to remotely deactivate a device

### PR DESCRIPTION
**Pull request by Walt-Intel**
## PR Checklist

- [ ] Unit Tests have been added for new changes
- [ ] API tests have been updated if applicable
- [ ] All commented code has been removed
- [ ] If you've added a dependency, you've ensured license is compatible with Apache 2.0 and clearly outlined the added dependency.

### CODE REVIEW

File: src/amt/DeviceAction.ts [hunk 1]
```diff
@@ -392,4 +392,12 @@ export class DeviceAction {
     logger.silly(`deleteAlarmClockOccurrences ${messages.COMPLETE}`)
     return deleteResponse.Envelope
   }
+
+  async unprovisionDevice (): Promise<Common.Models.Envelope<any>> {
```
Comment (matt-primrose) [inline src/amt/DeviceAction.ts:396]: @Walt-Intel Can you add the following test to the deviceAction.test.ts
`describe('unprovision amt', () => {
    it('should return response from AMT', async () => {
      sendSpy.mockResolvedValueOnce({Envelope:null})
      const result = await device.unprovisionDevice()
      expect(result).toBe(null)
    })
  })`

File: src/utils/SecretManagerService.ts [hunk 2]
```diff
@@ -81,6 +81,16 @@ export class SecretManagerService implements ISecretManagerService {
     return certs
   }
 
+  async deleteSecretAtPath (path: string): Promise<void> {
+    // to permanently delete the key, we use metadata path
+    const basePath = Environment.Config.secrets_path.replace('/data/', '/metadata/')
```
Comment (matt-primrose) [inline src/utils/SecretManagerService.ts:86]: @Walt-Intel can you add the following test to secretManagerService.test.ts:
`test('should remove device from secret store', async () => {
  const deleteSecretAtPathSpy = jest.spyOn(secretManagerService, 'deleteSecretAtPath').mockResolvedValue(null)
  const result = await secretManagerService.deleteSecretAtPath('test')
  expect(deleteSecretAtPathSpy).toHaveBeenCalledWith('test')
  expect(result).toBe(null)
})`

**Review by matt-primrose — CHANGES_REQUESTED** [review]
See comments that address code coverage gaps on changes

File: src/routes/amt/deactivate.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+export async function deactivate (req: Request, res: Response): Promise<void> {
+  try {
+    const guid: string = req.params.guid
+
+    const result = await req.deviceAction.unprovisionDevice()
+    if (result.Body?.Unprovision_OUTPUT?.ReturnValue === 0) {
+      await req.db.devices.delete(guid)
+      await req.secrets.deleteSecretAtPath(`devices/${guid}`)
+      res.status(200).json({ status: 'SUCCESS' })
```
Comment (rsdmike) [inline src/routes/amt/deactivate.ts:19]: Can you also add the removal of the device from memory like we talked about in planning. Something like
`delete devices[guid]`

**Review by rsdmike — CHANGES_REQUESTED** [review]

**Review by matt-primrose — APPROVED** [review]
Good from my point of view.  Implement @rsdmike comment and good to merge!

**Review by rsdmike — APPROVED** [review]

**Review by matt-primrose — DISMISSED** [review]

File: src/utils/secretManagerService.test.ts [hunk 4]
```diff
@@ -134,6 +135,18 @@ test('should return null if MPS certs does not exists', async () => {
...
+test('should remove device from secret store', async () => {
+  gotDeleteSpy = jest.spyOn(secretManagerService.gotClient, 'delete').mockImplementation(() => {
+    return {
+      json: jest.fn(() => {
+        return null
+      })
+    } as any
+  })
+  await secretManagerService.deleteSecretAtPath('test')
+  expect(gotDeleteSpy).toHaveBeenCalled()
```
Comment (rsdmike) [inline src/utils/secretManagerService.test.ts:147]: only thing i would change here is to ensure the call was made with what is expected. So:

``` suggestion
expect(gotDeleteSpy).toHaveBeenCalledWith('test',{ prefixUrl: 'theexpectedurl'})
```

**Review by matt-primrose — DISMISSED** [review]

**Review by madhavilosetty-intel — APPROVED** [review]

**Pull request closed without merging**